// @trpc/server/dist/adapters/node-http/content-type/json/getPostBody.mjs
import '../../../../unstable-core-do-not-import/rpc/codes.mjs';
import { TRPCError } from '../../../../unstable-core-do-not-import/error/TRPCError.mjs';
import '../../../../unstable-core-do-not-import/rootConfig.mjs';

// @trpc/server
async function getPostBody(opts) {
    const { req , maxBodySize =Infinity  } = opts;
    return new Promise((resolve)=>{
        if ('body' in req) {
            resolve({
                ok: true,
                data: req.body,
                // If the request headers specifies a content-type, we assume that the body has been preprocessed
                preprocessed: !!req.headers['content-type']?.startsWith('application/json')
            });
            return;
        }
        let body = '';
        let hasBody = false;
        req.on('data', function(data) {
            body += data;
            hasBody = true;
            if (body.length > maxBodySize) {
                resolve({
                    ok: false,
                    error: new TRPCError({
                        code: 'PAYLOAD_TOO_LARGE'
                    })
                });
            }
        });
        req.on('end', ()=>{
            resolve({
                ok: true,
                data: hasBody ? body : undefined,
                preprocessed: false
            });
        });
    });
}

export { getPostBody };
